### PR TITLE
TINY-6326: Now converts several columns at once.

### PR DESCRIPTION
**Pull request by HAFRMO**
Related Ticket:

Description of Changes:
We can now change one or more columns to headers.

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/` for new features (if applicable)
* [x] License headers added on new files (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 1]
```diff
@@ -305,8 +321,8 @@ export const splitCellIntoColumns = run(opSplitCellIntoColumns, onCell, resize,
 export const splitCellIntoRows = run(opSplitCellIntoRows, onCell, Fun.noop, Fun.noop, Generators.modification);
 export const eraseColumns = run(opEraseColumns, onCells, resize, prune, Generators.modification);
 export const eraseRows = run(opEraseRows, onCells, Fun.noop, prune, Generators.modification);
-export const makeColumnHeader = run(opMakeColumnHeader, onCell, Fun.noop, Fun.noop, Generators.transform('row', 'th'));
-export const unmakeColumnHeader = run(opUnmakeColumnHeader, onCell, Fun.noop, Fun.noop, Generators.transform(null, 'td'));
+export const makeColumnHeader = run(opMakeColumnHeader, onCells, Fun.noop, Fun.noop, Generators.transform('row', 'th'));
+export const unmakeColumnHeader = run(opUnmakeColumnHeader, onCells, Fun.noop, Fun.noop, Generators.transform(null, 'td'));
 export const makeRowHeader = run(opMakeRowHeader, onCell, Fun.noop, Fun.noop, Generators.transform('col', 'th'));
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:326]: Was there a reason this also wasn't changed, as it appears it has the same issue? If we're changing one, we should do both so as to keep things consistent.

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 2]
```diff
@@ -149,19 +149,35 @@ const opMakeRowHeader = function (grid: Structs.RowCells[], detail: Structs.Deta
   return bundle(newGrid, detail.row, detail.column);
 };
 
-const opMakeColumnHeader = function (grid: Structs.RowCells[], detail: Structs.DetailExt, comparator: CompElm, genWrappers: GeneratorsTransform) {
-  const newGrid = TransformOperations.replaceColumn(grid, detail.column, comparator, genWrappers.replaceOrInit);
-  return bundle(newGrid, detail.row, detail.column);
+const opMakeColumnHeader = function (initialGrid: Structs.RowCells[], details: Structs.DetailExt[], comparator: CompElm, genWrappers: GeneratorsTransform) {
+  const columns = uniqueColumns(details);
+  let currentGrid = initialGrid;
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:154]: No mutation please. Use `Arr.foldl` instead to avoid the mutation.

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 3]
```diff
@@ -149,19 +149,35 @@ const opMakeRowHeader = function (grid: Structs.RowCells[], detail: Structs.Deta
...
   const newGrid = TransformOperations.replaceRow(grid, detail.row, comparator, genWrappers.replaceOrInit);
   return bundle(newGrid, detail.row, detail.column);
 };
 
-const opUnmakeColumnHeader = function (grid: Structs.RowCells[], detail: Structs.DetailExt, comparator: CompElm, genWrappers: GeneratorsTransform) {
-  const newGrid = TransformOperations.replaceColumn(grid, detail.column, comparator, genWrappers.replaceOrInit);
-  return bundle(newGrid, detail.row, detail.column);
+const opUnmakeColumnHeader = function (initialGrid: Structs.RowCells[], details: Structs.DetailExt[], comparator: CompElm, genWrappers: GeneratorsTransform) {
+  const columns = uniqueColumns(details);
+  let currentGrid = initialGrid;
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:172]: Likewise, no mutation please.

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 4]
```diff
@@ -149,19 +149,35 @@ const opMakeRowHeader = function (grid: Structs.RowCells[], detail: Structs.Deta
...
+
+  Arr.each(columns, (column) => {
+    currentGrid = TransformOperations.replaceColumn(currentGrid, column.column, comparator, genWrappers.replaceOrInit);
+  });
+
+  const first = details[0];
+
+  return bundle(currentGrid, first ? first.row : 0, first ? first.column : 0);
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:162]: Strict null/undefined checks? as atm this is just a falsy check

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 5]
```diff
@@ -149,19 +149,35 @@ const opMakeRowHeader = function (grid: Structs.RowCells[], detail: Structs.Deta
...
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:180]: Strict null/undefined checks? as atm this is just a falsy check

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 6]
```diff
@@ -305,8 +321,8 @@ export const splitCellIntoColumns = run(opSplitCellIntoColumns, onCell, resize,
 export const splitCellIntoRows = run(opSplitCellIntoRows, onCell, Fun.noop, Fun.noop, Generators.modification);
 export const eraseColumns = run(opEraseColumns, onCells, resize, prune, Generators.modification);
 export const eraseRows = run(opEraseRows, onCells, Fun.noop, prune, Generators.modification);
-export const makeColumnHeader = run(opMakeColumnHeader, onCell, Fun.noop, Fun.noop, Generators.transform('row', 'th'));
-export const unmakeColumnHeader = run(opUnmakeColumnHeader, onCell, Fun.noop, Fun.noop, Generators.transform(null, 'td'));
+export const makeColumnHeader = run(opMakeColumnHeader, onCells, Fun.noop, Fun.noop, Generators.transform('row', 'th'));
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:324]: This is technically a breaking API change, so please be sure to include a major version bump for snooker in `versions.txt`.

File: modules/snooker/src/test/ts/module/ephox/snooker/test/Assertions.ts [hunk 7]
```diff
@@ -9,18 +9,31 @@ import { RunOperationOutput, TargetElement, TargetPasteRows, TargetSelection } f
...
 
+type OperationMultile<T> = (
+  wire: ResizeWire,
+  table: SugarElement,
+  target: T,
+  generators: Generators,
+) => Optional<RunOperationOutput>;
+
+interface TargetLocation {
+  section: number;
```
Comment (lnewson) [inline modules/snooker/src/test/ts/module/ephox/snooker/test/Assertions.ts:27]: readonly properties if possible please?

File: modules/snooker/src/test/ts/module/ephox/snooker/test/Assertions.ts [hunk 8]
```diff
@@ -9,18 +9,31 @@ import { RunOperationOutput, TargetElement, TargetPasteRows, TargetSelection } f
...
 
-type Op<T> = (
+type OperationSingle<T> = (
   wire: ResizeWire,
   table: SugarElement,
   target: T,
   generators: Generators,
 ) => Optional<RunOperationOutput>;
 
+type OperationMultile<T> = (
```
Comment (lnewson) [inline modules/snooker/src/test/ts/module/ephox/snooker/test/Assertions.ts:19]: `Multile` isn't a word, so maybe `Multi` or `Multiple`?

File: modules/snooker/src/test/ts/module/ephox/snooker/test/Assertions.ts [hunk 9]
```diff
@@ -47,11 +60,44 @@ const checkOld = (
...
+        Hierarchy.follow(table, [ path.section, path.row, path.column, 0 ]).getOrDie()
+      )
+    },
+    Bridge.generators
+  );
+
+  const actualPath = Hierarchy.path(table, result.getOrDie().cursor.getOrDie()).getOrDie('could not find path');
+  assert.eq([ expCell.section, expCell.row, expCell.column ], actualPath);
+
+  // Presence.assertHas(expected, table, 'checking the operation on table: ' + Html.getOuter(table));
```
Comment (lnewson) [inline modules/snooker/src/test/ts/module/ephox/snooker/test/Assertions.ts:85]: Was there a reason this is commented out? If it's not useful, then should it just be deleted?

File: modules/tinymce/src/plugins/table/main/ts/actions/TableActions.ts [hunk 10]
```diff
@@ -42,8 +42,8 @@ export interface TableActions {
   pasteRowsAfter: AdvancedPasteTableAction;
   setTableCellType: SimpleTableAction;
   setTableRowType: SimpleTableAction;
-  makeColumnHeader: ElementTableAction;
-  unmakeColumnHeader: ElementTableAction;
+  makeColumnHeader: CombinedTargetsTableAction;
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/main/ts/actions/TableActions.ts:45]: Do we have an automated test that covers this change? If not, then can you please add one?

**Review by lnewson — CHANGES_REQUESTED** [review]

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts (hunk 6 above)
Comment (lnewson) [reply modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:324]: Also this name is now misleading, as we're operating on multiple columns. So instead of a breaking change, maybe a new function should have been added instead (so as to keep consistent with `insertColumnBefore` vs `insertColumnsBefore`)?

File: modules/snooker/src/test/ts/module/ephox/snooker/test/Assertions.ts (hunk 8 above)
Comment (HAFRMO) [reply modules/snooker/src/test/ts/module/ephox/snooker/test/Assertions.ts:19]: Well, it was an attempt to spell `multiple`...

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts (hunk 1 above)
Comment (HAFRMO) [reply modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:326]: What is 'this' in this situation? `makeRowHeader`?

Comment (lnewson) [reply modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:326]: Yes, that's the line I've commented on. Sorry it wasn't clear.

Comment (lnewson) [reply modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:326]: There's also `unmakeRowHeader` below that this also applies to sorry.

File: modules/snooker/src/test/ts/module/ephox/snooker/test/Assertions.ts (hunk 9 above)
Comment (HAFRMO) [reply modules/snooker/src/test/ts/module/ephox/snooker/test/Assertions.ts:85]: Because it tagged along from `checkOld`, and I forgot to remove it.

File: modules/tinymce/src/plugins/table/main/ts/actions/TableActions.ts (hunk 10 above)
Comment (HAFRMO) [reply modules/tinymce/src/plugins/table/main/ts/actions/TableActions.ts:45]: Which change do you mean exactly?

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts (hunk 1 above)
Comment (HAFRMO) [reply modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:326]: Wasn't really mentioned in the task, so I wasn't sure if I should change it or not.
But I'll make the change.

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts (hunk 4 above)
Comment (HAFRMO) [reply modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:162]: I thought the truthy checks would be suitable. I can change them into `Type.isUndefined` or `Type.isObject` if that'll read better.

Comment (lnewson) [reply modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:162]: Sorry, you're right it's a truthy check, not a falsy check. Anyways, truthy/falsy checks should only be done for boolean variables. So yes, this should either use an `Optional` or use an explicit `Type` check.

My bigger question reading more now though, is why do you even need that check? It should never be an empty array from what I can see, as `onCells` guarantees it'll always have at least one value. You'll see everywhere else it's used, there's no check like this. eg: from `opInsertRowsBefore`:
```
  const example = details[0].row;
  const targetIndex = details[0].row;
```

Comment (HAFRMO) [reply modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts]: It shouldn't, but I also felt that it'd be poorly done to assume that there was something in there, since the input is from outside of the module and/or plugin.

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 11]
```diff
@@ -149,21 +149,65 @@ const opMakeRowHeader = function (grid: Structs.RowCells[], detail: Structs.Deta
   return bundle(newGrid, detail.row, detail.column);
 };
 
-const opMakeColumnHeader = function (grid: Structs.RowCells[], detail: Structs.DetailExt, comparator: CompElm, genWrappers: GeneratorsTransform) {
-  const newGrid = TransformOperations.replaceColumn(grid, detail.column, comparator, genWrappers.replaceOrInit);
+// Function added to match opMakeColumnsHeader, but not used anywhere.
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:152]: This isn't true, it's available in the public API. While it's not currently used anywhere, it very well could be and then this comment would be very misleading. So I'm going to ask that you remove this please, as it's the type of thing that while accurate now will degrade over time.

- [ ] Remove the 2 comments like this

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 12]
```diff
@@ -149,21 +149,65 @@ const opMakeRowHeader = function (grid: Structs.RowCells[], detail: Structs.Deta
...
+const opMakeRowsHeader = function (initialGrid: Structs.RowCells[], details: Structs.DetailExt[], comparator: CompElm, genWrappers: GeneratorsTransform) {
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:153]: Arrow functions please

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 13]
```diff
@@ -149,21 +149,65 @@ const opMakeRowHeader = function (grid: Structs.RowCells[], detail: Structs.Deta
...
+  return bundle(newGrid, details[0].row, details[0].column);
+};
+
+const opMakeColumnHeader = function (initialGrid: Structs.RowCells[], detail: Structs.DetailExt, comparator: CompElm, genWrappers: GeneratorsTransform) {
+  const newGrid = TransformOperations.replaceColumn(initialGrid, detail.column, comparator, genWrappers.replaceOrInit);
+
   return bundle(newGrid, detail.row, detail.column);
 };
 
+const opMakeColumnsHeader = function (initialGrid: Structs.RowCells[], details: Structs.DetailExt[], comparator: CompElm, genWrappers: GeneratorsTransform) {
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:169]: Arrow functions please

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 14]
```diff
@@ -149,21 +149,65 @@ const opMakeRowHeader = function (grid: Structs.RowCells[], detail: Structs.Deta
...
+
 const opUnmakeRowHeader = function (grid: Structs.RowCells[], detail: Structs.DetailExt, comparator: CompElm, genWrappers: GeneratorsTransform) {
   const newGrid = TransformOperations.replaceRow(grid, detail.row, comparator, genWrappers.replaceOrInit);
   return bundle(newGrid, detail.row, detail.column);
 };
 
-const opUnmakeColumnHeader = function (grid: Structs.RowCells[], detail: Structs.DetailExt, comparator: CompElm, genWrappers: GeneratorsTransform) {
-  const newGrid = TransformOperations.replaceColumn(grid, detail.column, comparator, genWrappers.replaceOrInit);
+// Function added to match opUnmakeColumnsHeader, but not used anywhere.
+const opUnmakeRowsHeader = function (initialGrid: Structs.RowCells[], details: Structs.DetailExt[], comparator: CompElm, genWrappers: GeneratorsTransform) {
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:185]: Arrow functions please

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 15]
```diff
@@ -149,21 +149,65 @@ const opMakeRowHeader = function (grid: Structs.RowCells[], detail: Structs.Deta
...
+  return bundle(newGrid, details[0].row, details[0].column);
+};
+
+const opUnmakeColumnHeader = function (initialGrid: Structs.RowCells[], detail: Structs.DetailExt, comparator: CompElm, genWrappers: GeneratorsTransform) {
+  const newGrid = TransformOperations.replaceColumn(initialGrid, detail.column, comparator, genWrappers.replaceOrInit);
+
   return bundle(newGrid, detail.row, detail.column);
 };
 
+const opUnmakeColumnsHeader = function (initialGrid: Structs.RowCells[], details: Structs.DetailExt[], comparator: CompElm, genWrappers: GeneratorsTransform) {
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:201]: Arrow functions please

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 16]
```diff
@@ -149,21 +149,65 @@ const opMakeRowHeader = function (grid: Structs.RowCells[], detail: Structs.Deta
...
 
-const opMakeColumnHeader = function (grid: Structs.RowCells[], detail: Structs.DetailExt, comparator: CompElm, genWrappers: GeneratorsTransform) {
-  const newGrid = TransformOperations.replaceColumn(grid, detail.column, comparator, genWrappers.replaceOrInit);
+// Function added to match opMakeColumnsHeader, but not used anywhere.
+const opMakeRowsHeader = function (initialGrid: Structs.RowCells[], details: Structs.DetailExt[], comparator: CompElm, genWrappers: GeneratorsTransform) {
+  const rows = uniqueRows(details);
+
+  const newGrid = Arr.foldl(rows, (currentGrid, row) =>
+    TransformOperations.replaceRow(currentGrid, row.row, comparator, genWrappers.replaceOrInit),
+  initialGrid);
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:158]: This is pretty hard to read due to the mixture of the fold and the function wrapping over multiple lines. Maybe something like this instead?
```suggestion
  const replacer = (currentGrid: Structs.RowCells, row: number) => TransformOperations.replaceRow(currentGrid, row.row, comparator, genWrappers.replaceOrInit);
  const newGrid = Arr.foldl(rows, replacer, initialGrid);
```

If you do change this, please replace the 3 other places that are the same as this.

- [ ] Consider using different formatting to make this more readable

File: modules/snooker/src/test/ts/module/ephox/snooker/test/Assertions.ts (hunk 8 above)
Comment (lnewson) [reply modules/snooker/src/test/ts/module/ephox/snooker/test/Assertions.ts:19]: This still hasn't been fixed, so can't mark as resolved sorry @HAFRMO

**Review by lnewson — APPROVED** [review]

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts (hunk 6 above)
Comment (lnewson) [reply modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts]: `versions.txt` still hasn't been updated to reflect this needs a minor version bump.

- [ ] Add snooker minor version bump to `versions.txt`

File: modules/tinymce/src/plugins/table/main/ts/actions/TableActions.ts (hunk 10 above)
Comment (lnewson) [reply modules/tinymce/src/plugins/table/main/ts/actions/TableActions.ts]: I could have sworn I replied to this... anyways the change is that you've changed what data needs to be passed to Snooker now from within the table plugin. So we need to make sure we have a test that covers making a selection across multiple columns a header. I know you have the snooker tests, but they are unit test, not functional tests. 

- [ ] Make sure we have a test in the table plugin that covers the changes made

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts (hunk 11 above)
Comment (metricjs) [reply modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:152]: Ah, I mentioned to Hannes about adding the comments tbh. At the time we were discussing how some (at least one?) method in the current API isn't used anywhere, which was a little confusing. I get what you mean about it degrading though... Hm.

File: modules/snooker/src/test/ts/browser/HeaderOperationsTest.ts [hunk 17]
```diff
@@ -3,278 +3,828 @@ import * as TableOperations from 'ephox/snooker/api/TableOperations';
 import * as Assertions from 'ephox/snooker/test/Assertions';
 
 UnitTest.test('HeaderOperationsTest', function () {
```
Comment (metricjs) [inline modules/snooker/src/test/ts/browser/HeaderOperationsTest.ts:5]: This file is _long_. I'm undecided on whether it's better as one file or if we should break it into multiple. I guess, if anyone else chimes in they can tie-break, I don't mind if it's left as is for now.

**Review by metricjs — APPROVED** [review]
Didn't see anything that doesn't already have comments :)

File: modules/tinymce/src/plugins/table/test/ts/browser/TableSectionApiTest.ts [hunk 18]
```diff
@@ -141,6 +173,25 @@ UnitTest.asynctest('browser.tinymce.plugins.table.TableSectionApiTest', (success
       ApiChains.cAssertContent(expectedContent)
     ]));
 
+  const cSwitchMultipleColumnsType = (startContent: string, expectedContent: string, command: string, type: string) =>
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/TableSectionApiTest.ts:176]: The `type` here could be better constrained to `'td' | `'th'`. That would then remove the need for the `searchingForType` variable below.

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 19]
```diff
@@ -36,191 +36,230 @@ const outcome = (grid: Structs.RowCells[], cursor: Optional<SugarElement>): Tabl
...
-    return Arr.exists(rest, function (currentDetail) {
-      return currentDetail.column === detail.column;
-    }) ? rest : rest.concat([ detail ]);
-  }, [] as Structs.DetailExt[]).sort(function (detailA, detailB) {
-    return detailA.column - detailB.column;
-  });
-};
-
-const opInsertRowBefore = function (grid: Structs.RowCells[], detail: Structs.DetailExt, comparator: CompElm, genWrappers: GeneratorsModification) {
+const uniqueRows = (details: Structs.DetailExt[]) =>
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:58]: This is very hard to read now (see my previous points about `foldl` readability) and I have to ask why was this changed?

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 20]
```diff
@@ -36,191 +36,230 @@ const outcome = (grid: Structs.RowCells[], cursor: Optional<SugarElement>): Tabl
...
+  Arr.foldl(details, (rest, detail) =>
+    Arr.exists(rest, (currentDetail) =>
+      currentDetail.row === detail.row
+    ) ? rest : rest.concat([ detail ]),
+  [] as Structs.DetailExt[]).sort((detailA, detailB) =>
+    detailA.row - detailB.row
+  );
+
+const uniqueColumns = (details: Structs.DetailExt[]) =>
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:67]: Likewise this is harder to read now.

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 21]
```diff
@@ -36,191 +36,230 @@ const outcome = (grid: Structs.RowCells[], cursor: Optional<SugarElement>): Tabl
...
 
-const opInsertRowsAfter = function (grid: Structs.RowCells[], details: Structs.DetailExt[], comparator: CompElm, genWrappers: GeneratorsModification) {
+const opInsertRowsAfter = (grid: Structs.RowCells[], details: Structs.DetailExt[], comparator: CompElm, genWrappers: GeneratorsModification) => {
   const rows = uniqueRows(details);
   const example = rows[rows.length - 1].row;
   const targetIndex = rows[rows.length - 1].row + rows[rows.length - 1].rowspan;
-  const newGrid = Arr.foldl(rows, function (newG, _row) {
-    return ModificationOperations.insertRowAt(newG, targetIndex, example, comparator, genWrappers.getOrInit);
-  }, grid);
+  const newGrid = Arr.foldl(rows, (newG, _row) =>
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:104]: Another place that's harder to read now.

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 22]
```diff
@@ -36,191 +36,230 @@ const outcome = (grid: Structs.RowCells[], cursor: Optional<SugarElement>): Tabl
...
 
-const opInsertRowsBefore = function (grid: Structs.RowCells[], details: Structs.DetailExt[], comparator: CompElm, genWrappers: GeneratorsModification) {
+const opInsertRowsBefore = (grid: Structs.RowCells[], details: Structs.DetailExt[], comparator: CompElm, genWrappers: GeneratorsModification) => {
   const example = details[0].row;
   const targetIndex = details[0].row;
   const rows = uniqueRows(details);
-  const newGrid = Arr.foldl(rows, function (newG, _row) {
-    return ModificationOperations.insertRowAt(newG, targetIndex, example, comparator, genWrappers.getOrInit);
-  }, grid);
+  const newGrid = Arr.foldl(rows, (newG, _row) =>
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:87]: Another place that's harder to read now.

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 23]
```diff
@@ -36,191 +36,230 @@ const outcome = (grid: Structs.RowCells[], cursor: Optional<SugarElement>): Tabl
...
 
-const opInsertColumnsBefore = function (grid: Structs.RowCells[], details: Structs.DetailExt[], comparator: CompElm, genWrappers: GeneratorsModification) {
+const opInsertColumnsBefore = (grid: Structs.RowCells[], details: Structs.DetailExt[], comparator: CompElm, genWrappers: GeneratorsModification) => {
   const columns = uniqueColumns(details);
   const example = columns[0].column;
   const targetIndex = columns[0].column;
-  const newGrid = Arr.foldl(columns, function (newG, _row) {
-    return ModificationOperations.insertColumnAt(newG, targetIndex, example, comparator, genWrappers.getOrInit);
-  }, grid);
+  const newGrid = Arr.foldl(columns, (newG, _row) =>
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:121]: Another place that's harder to read now.

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 24]
```diff
@@ -36,191 +36,230 @@ const outcome = (grid: Structs.RowCells[], cursor: Optional<SugarElement>): Tabl
...
 
-const opInsertColumnsAfter = function (grid: Structs.RowCells[], details: Structs.DetailExt[], comparator: CompElm, genWrappers: GeneratorsModification) {
+const opInsertColumnsAfter = (grid: Structs.RowCells[], details: Structs.DetailExt[], comparator: CompElm, genWrappers: GeneratorsModification) => {
   const example = details[details.length - 1].column;
   const targetIndex = details[details.length - 1].column + details[details.length - 1].colspan;
   const columns = uniqueColumns(details);
-  const newGrid = Arr.foldl(columns, function (newG, _row) {
-    return ModificationOperations.insertColumnAt(newG, targetIndex, example, comparator, genWrappers.getOrInit);
-  }, grid);
+  const newGrid = Arr.foldl(columns, (newG, _row) =>
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:138]: Another place that's harder to read now.

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 25]
```diff
@@ -36,191 +36,230 @@ const outcome = (grid: Structs.RowCells[], cursor: Optional<SugarElement>): Tabl
...
   const newGrid = MergingOperations.merge(grid, mergable.bounds, comparator, Fun.constant(cells[0]));
   return outcome(newGrid, Optional.from(cells[0]));
 };
 
-const opUnmergeCells = function (grid: Structs.RowCells[], unmergable: SugarElement[], comparator: CompElm, genWrappers: GeneratorsMerging) {
-  const newGrid = Arr.foldr(unmergable, function (b, cell) {
-    return MergingOperations.unmerge(b, cell, comparator, genWrappers.combine(cell));
-  }, grid);
+const opUnmergeCells = (grid: Structs.RowCells[], unmergable: SugarElement[], comparator: CompElm, genWrappers: GeneratorsMerging) => {
+  const newGrid = Arr.foldr(unmergable, (b, cell) =>
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:240]: Another place that's harder to read now.

**Review by lnewson — CHANGES_REQUESTED** [review]
Marking as changes requested, as again you've avoided Sugar. You need to remember that should be used as a first preference in most cases. Also, you've introduced readability issues again, that I had flagged in a previous review.

File: modules/tinymce/src/plugins/table/test/ts/browser/TableSectionApiTest.ts [hunk 26]
```diff
@@ -141,6 +173,25 @@ UnitTest.asynctest('browser.tinymce.plugins.table.TableSectionApiTest', (success
...
+    Log.chain('TINY-6326', `Switch to ${type}, command = ${command}`, Chain.fromParent(Chain.identity, [
+      ApiChains.cSetContent(startContent),
+      Chain.op((editor: Editor) => {
+        const searchingForType = type === 'th' ? 'td' : 'th';
+
+        const getCells = (table): HTMLTableCellElement[] =>
```
[the remark this answers is not in the document]
Comment (HAFRMO) [reply modules/tinymce/src/plugins/table/test/ts/browser/TableSectionApiTest.ts:182]: Me and Millie were checking how to do a selection of table-cells for the tests, and this way was what we found. It seemed good enough, but I'll try to conver it to sugar.

File: modules/tinymce/src/plugins/table/test/ts/browser/TableSectionApiTest.ts (hunk 18 above)
Comment (HAFRMO) [reply modules/tinymce/src/plugins/table/test/ts/browser/TableSectionApiTest.ts:176]: No, you misunderstand the use for the `searchingForType`. It is the opposite of `type`, as `type` is what it will be changed to, `searchingForType` is what it currently is.
I'll type-restict `type` though.

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts (hunk 19 above)
Comment (HAFRMO) [reply modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:58]: Because putting several functions inside of each other on the same line is really hard to read. The newlines are, admittedly, not the easiest, but at least they provide some division between the different functions, as opposed to simply lumping them all onto one big pile. Brackets would make it much easier to read, but we keep arrow-functions naked when possible.

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 27]
```diff
@@ -36,191 +36,228 @@ const outcome = (grid: Structs.RowCells[], cursor: Optional<SugarElement>): Tabl
...
-  return Arr.foldl(details, function (rest, detail) {
-    return Arr.exists(rest, function (currentDetail) {
-      return currentDetail.column === detail.column;
-    }) ? rest : rest.concat([ detail ]);
-  }, [] as Structs.DetailExt[]).sort(function (detailA, detailB) {
-    return detailA.column - detailB.column;
-  });
+const uniqueColumns = (details: Structs.DetailExt[]) => {
+  const uniqueCheck = (rest: Structs.DetailExt[], detail: Structs.DetailExt) => {
+    const colunExists = Arr.exists(rest, (currentDetail) => currentDetail.column === detail.column);
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:72]: There's a typo in `columnExists` here (missing `m`).

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts (hunk 25 above)
Comment (lnewson) [reply modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:238]: This still needs to be fixed

- [ ] Fix formatting

File: modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts [hunk 28]
```diff
@@ -36,191 +36,228 @@ const outcome = (grid: Structs.RowCells[], cursor: Optional<SugarElement>): Tabl
...
   const mergedGrid = TableMerge.merge(startAddress, grid, gridB, pasteDetails.generators, comparator);
-  return mergedGrid.fold(function () {
-    return outcome(grid, Optional.some(pasteDetails.element));
-  }, function (nuGrid) {
+  return mergedGrid.fold(() =>
+    outcome(grid, Optional.some(pasteDetails.element)),
+  (nuGrid) => {
     const cursor = elementFromGrid(nuGrid, pasteDetails.row, pasteDetails.column);
     return outcome(nuGrid, cursor);
   });
```
Comment (lnewson) [inline modules/snooker/src/main/ts/ephox/snooker/api/TableOperations.ts:220]: This also has formatting issues that need to be fixed
```suggestion
  return mergedGrid.fold(
    () => outcome(grid, Optional.some(pasteDetails.element)),
    (nuGrid) => {
      const cursor = elementFromGrid(nuGrid, pasteDetails.row, pasteDetails.column);
      return outcome(nuGrid, cursor);
    }
  );
```

- [ ] Fix formatting

File: modules/tinymce/src/plugins/table/test/ts/browser/TableSectionApiTest.ts [hunk 29]
```diff
@@ -130,6 +156,21 @@ UnitTest.asynctest('browser.tinymce.plugins.table.TableSectionApiTest', (success
 </tbody>
 </table>`;
 
+  const selectAllCells = (type: 'td' | 'th') =>
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/TableSectionApiTest.ts:159]: This is a chain, but it's missing the naming convention
```suggestion
  const cSelectAllCells = (type: 'td' | 'th') =>
```

- [ ] Fix function name

**Review by lnewson — APPROVED** [review]
LGTM, just a few more formatting issues that still haven't been fixed yet.

**Review by ltrouton — APPROVED** [review]
